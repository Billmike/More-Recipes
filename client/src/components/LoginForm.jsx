import React, { Component } from 'react';
import validateInput from '../../../server/validators/validatesignin';
import '../assets/css/signup.css';

class LoginForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      email: '',
      password: '',
      errors: {}
    };
    this.onEmailChange = this.onEmailChange.bind(this);
    this.onPasswordChange = this.onPasswordChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  };

  onEmailChange(event) {
    this.setState({ [event.target.name]: event.target.value });
  }

  onPasswordChange(event) {
    this.setState({ [event.target.name]: event.target.value });
  }

  onSubmit(event) {
    event.preventDefault();
    this.setState({ errors: {}, isLoading: true });
    if(this.isValid()) {
      this.props.signinRequest(this.state)
        .then(() => {
          this.props.history.push('/dashboard');
        })
        .catch((error) => {
          this.setState({ isLoading: false });
          if (error === 'Invalid email or password.') {
            return toastr.error('Invalid login details');
          }
        });
    }
  }

  isValid() {
    const { errors, valid } = validateInput(this.state);

    if(!valid) {
      this.setState({ errors });
    }

    return valid;
  }

  render() {
    const { errors } = this.state;
    return (
      <form onSubmit={this.onSubmit}>
        <div id="login-box">
          <div className="left">
            <h1>Sign In</h1>
            <div>
              <input
                type="text"
                name="email"
                value={this.state.email}
                onChange={this.onEmailChange}
                placeholder="E-mail"
              />
              {errors.email && <span className="help-block red-errors">{errors.email}</span>}
            </div>
            <div>
              <input
                type="password"
                name="password"
                value={this.state.password}
                onChange={this.onPasswordChange}
                placeholder="Password"
              />
              {errors.password && <span className="help-block red-errors">{errors.password}</span>}
            </div>
            <input
              disabled={this.state.isLoading}
              type="submit"
              name="signup_submit"
              value="Sign In"
            />
          </div>

          <div className="right">
            <span className="loginwith">
              Sign in with<br />social network
            </span>

            <button className="social-signin facebook">
              Log in with facebook
            </button>
            <button className="social-signin twitter">
              Log in with Twitter
            </button>
            <button className="social-signin google">
              Log in with Google+
            </button>
          </div>
          <div className="or">OR</div>
        </div>
      </form>
    );
  }
}

export default LoginForm;
