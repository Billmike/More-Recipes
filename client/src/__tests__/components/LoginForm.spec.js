import React from 'react';
import { shallow } from 'enzyme';
import LoginForm from '../../components/LoginForm';

describe('<LoginForm />', () => {
  it('Should render the form correctly', () => {
    const wrapper = shallow(<LoginForm />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('div').length).toBe(6);
  });
  it('Should handle input change', () => {
    const value = 'randomemail@gmail.com';
    const wrapper = shallow(<LoginForm />);
    wrapper.find('#email').simulate('change', {
      target: { name: 'email', value }
    });
    expect(wrapper.state('email')).toBe(value);
  });
  it('Should populate the form field and update the state', () => {
    const signinRequest = jest.fn();
    const wrapper = shallow(<LoginForm signinRequest={() => Promise.resolve()} />);
    wrapper.instance().setState({
      email: 'randomemail@gmail.com',
      password: 'qwertyuiop',
      errors: {}
    });
    wrapper.update();
    wrapper.find('form').simulate('submit', {
      preventDefault: jest.fn()
    });
    expect(wrapper.instance().state).toEqual({
      email: 'randomemail@gmail.com',
      password: 'qwertyuiop',
      errors: {}
    });
  });
  it('Should render errors on invalid form submission', () => {
    const wrapper = shallow(<LoginForm />);
    wrapper.find('form').simulate('submit', {
      preventDefault: jest.fn()
    });
    expect(wrapper.state('errors')).toBeTruthy();
  });
});
