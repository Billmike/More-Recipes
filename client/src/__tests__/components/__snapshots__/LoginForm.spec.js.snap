// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginForm /> Should render the form correctly 1`] = `
<div
  className="container"
  id="loginContainer"
>
  <div
    className="card text-center card-form has-feedback"
  >
    <div
      className="card-body"
    >
      <form
        id="loginFormBody"
        onSubmit={[Function]}
      >
        <h1
          className="sign-up-h3"
        >
          Login
        </h1>
        <div
          className="form-group"
        >
          <input
            autoComplete="email"
            className="form-control form-control-lg"
            id="email"
            name="email"
            onChange={[Function]}
            placeholder="Email Address"
            type="email"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            autoComplete="current-password"
            className="form-control form-control-lg"
            id="password"
            name="password"
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <div />
        <input
          className="btn submit-btn size-bt btn-block"
          id="submitButton"
          type="submit"
          value="Submit"
        />
      </form>
      <p
        className="signin-pad"
      >
         Not yet registered?
        <Link
          className="account"
          replace={false}
          to="/register"
        >
          Let's hook you up with an account.
        </Link>
      </p>
    </div>
  </div>
</div>
`;
